File: Engine/ceCore/resource/javafile.cc
#include <ceCore/resource/javafile.hh>
#include <ceCore/java.hh>
#include <ceCore/java_methods.hh>

namespace ce
{

static std::string s_name = "";

const std::string &JavaFile::GetName() const
{
  static JavaCallObject<jstring>
          getName(Java::Get(), this, "org/crimsonedge/core/resource/JavaFile", "getName", JAVA_STRING);
  jstring res = getName.call(Java::Get(), 0);
  if (res)
  {
    m_name = Java::Get()->GetStringUTFChars(res, 0);
  }
  return m_name;
}
bool JavaFile::Open(eAccessMode accessMode, eOpenMode openMode)
{
  static JavaCallBoolean2<jint, jint>
      open(Java::Get(), this, "org/crimsonedge/core/resource/JavaFile", "open", JAVA_INT, JAVA_INT);
  return open.call(Java::Get(), accessMode, openMode, false);
}
bool JavaFile::IsOpen() const
{
  static JavaCallBoolean isOpen(Java::Get(), this, "org/crimsonedge/core/resource/JavaFile", "isOpen");
  return isOpen.call(Java::Get(), false);
}

bool JavaFile::IsEOF() const
{
  static JavaCallBoolean isOpen(Java::Get(), this, "org/crimsonedge/core/resource/JavaFile", "isEOF");
  return isOpen.call(Java::Get(), false);

}
void JavaFile::Seek(eSeekMode mode, long offset)
{
  static JavaCallVoid2<jint, jlong>
      seek(Java::Get(), this, "org/crimsonedge/core/resource/JavaFile", "seek", JAVA_INT, JAVA_LONG);
  seek.call(Java::Get(), mode, offset);
}


long JavaFile::Tell() const
{
  static JavaCallLong tell(Java::Get(), this, "org/crimsonedge/core/resource/JavaFile", "tell");
  return (long) tell.call(Java::Get(), 0);
}

Size JavaFile::Read(Size elementSize, Size elementCount, void *buffer)
{
  size_t     size   = elementSize * elementCount;
  JNIEnv     *pEnv  = Java::Get();
  jbyteArray pArray = pEnv->NewByteArray((jsize) size);

  static JavaCallInt3<jint, jint, jbyteArray> read(pEnv,
                                                   this,
                                                   "org/crimsonedge/core/resource/JavaFile",
                                                   "read",
                                                   JAVA_INT,
                                                   JAVA_INT,
                                                   JAVA_BYTE_ARRAY);

  jint readBytes = read.call(pEnv, elementSize, elementCount, pArray, 0);

  jbyte *bytes = pEnv->GetByteArrayElements(pArray, 0);
  memcpy(buffer, bytes, readBytes);
  pEnv->ReleaseByteArrayElements(pArray, bytes, 0);


  pEnv->DeleteLocalRef(pArray);

  return readBytes;
}
Size JavaFile::Write(Size elementSize, Size elementCount, const void *buffer)
{
  return 0;
}
void JavaFile::Close()
{
  static JavaCallVoid close(Java::Get(), this, "org/crimsonedge/core/resource/JavaFile", "close");
  close.call(Java::Get());
}

}

File: JCryoStorm/CeCore/src/main/java/org/crimsonedge/core/resource/JavaFile.java
package org.crimsonedge.core.resource;

import org.crimsonedge.core.CeClass;
import org.crimsonedge.core.CoreObject;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.nio.ByteBuffer;

@CeClass("cs::csJavaFile")
public class csJavaFile extends CoreObject implements IFile {

    private final String name;

    private final byte[] data;

    private final InputStream inputStream;

    private int ptr;

    public csJavaFile(String name, InputStream inputStream) {
        super();
        this.name = name;
        this.inputStream = inputStream;
        this.data = readAll(inputStream);
        this.ptr = 0;
    }

    @Override
    public String getName() {
        return this.name;
    }

    public boolean open(int accessMode, int openMode) {
        return open(EAccessMode.values()[accessMode],
                    EOpenMode.values()[openMode]);
    }

    @Override
    public boolean open(EAccessMode accessMode, EOpenMode openMode) {
        return this.data != null && this.inputStream != null;
    }

    @Override
    public boolean isOpen() {
        return this.data != null && this.inputStream != null;
    }

    @Override
    public boolean isEOF() {
        return this.ptr >= this.data.length;
    }

    public void seek(int seekMode, long position) {
        seek(ESeekMode.values()[seekMode], position);
    }

    @Override
    public void seek(ESeekMode seekMode, long position) {
        switch (seekMode) {
            case SET:
                this.ptr = (int)position;
                break;
            case CUR:
                this.ptr += position;
                break;
            case END:
                this.ptr = this.data.length - (int)position;
                break;
        }
    }

    @Override
    public long tell() {
        return this.ptr;
    }

    @Override
    public int read(int elementSize, int elementCount, byte[] byteBuffer) {
        int numBytes = elementCount * elementSize;
        int remaining = this.data.length - this.ptr;
        int bytesRead = Math.min(numBytes, remaining);


        for (int i = 0; i < bytesRead; i++) {
            byteBuffer[i] = this.data[this.ptr++];
        }

        return bytesRead;
    }

    @Override
    public int write(int elementSize, int elementCount, ByteBuffer byteBuffer) {
        return 0;
    }

    @Override
    public void close() {
        try {
            this.inputStream.close();
        }
        catch (Exception e) {
            //
        }
    }


    private static byte[] readAll(InputStream inputStream) {
        try {
            // Create a ByteArrayOutputStream
            ByteArrayOutputStream buffer = new ByteArrayOutputStream();

            // Create a byte array for the buffer
            int nRead;
            byte[] data = new byte[1024];

            // Read the bytes from the input stream
            while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
                buffer.write(data, 0, nRead);
            }

            // Get the byte array
            return buffer.toByteArray();
        }
        catch (Exception e) {
            return null;
        }
    }
}
